src/solvers/smt2/smt2_parser.cpp

```diff
@@ -438,11 +438,8 @@ exprt smt2_parsert::function_application_fp(const exprt::operandst &op)
   const auto width_f = to_unsignedbv_type(op[2].type()).get_width();
 
   // stitch the bits together
-  concatenation_exprt c(bv_typet(width_e + width_f + 1));
-  c.operands() = {op[0], op[1], op[2]};
-
-  return std::move(
-    typecast_exprt(c, ieee_float_spect(width_f, width_e).to_type()));
+  return concatenation_exprt(
+    exprt::operandst(op), ieee_float_spect(width_f, width_e).to_type());
 }
```

Comment:
There's an implicit type conversion now happening here, as part of the conversion.
That will work with our flattening-backend, but will fail with the SMT2 backend, where these conversions need to be casts.

Reply:
Ok, good to know, I'll fix this - but I was guided by other places doing the same, so I will actually fix it in multiple places.

Reply:
I have added a second commit that is the result of reviewing all uses of `concatenation_exprt`.

Reply:
Apologies, I should have been clearer. Unfortunately, the meaning of typecast_exprt is complicated. It does a semantic cast, i.e., it attempts to preserve the _interpretation_ of the expression. The current code constructs an unisignedbv_typet expression, which is then casted to floatbv. This means that say 0u becomes 0.0f; what is needed here is a reinterpretation, i.e., a cast that doesn't change the representation.

In an ideal world, I think these two casts should be two separate expressions, which would make it clear whether reinterpretation happens or an actual transformation. In our current set-up, the way to do the reinterpretation is to first produce our bv_typet(), which has no meaning, and then cast that to the floating-point type.

Reply:
I have pushed another update, now using `bv_typet`. Is that now correct? Note that I have made this change in two places.